failure in host/test_ucp_mem_type_alloc_before_init.xfer/0 with UCX_RNDV_SCHEME=put_zcopy under valgrind

```
$ UCX_RNDV_SCHEME=put_zcopy GTEST_FILTER=host/test_ucp_mem_type_alloc_before_init.xfer/0 make test_valgrind
env LD_LIBRARY_PATH="/usr/lib64/mlnx_ofed/valgrind::" \
stdbuf -e0 -o0 valgrind --tool=memcheck --leak-check=full --track-origins=yes --fair-sched=try --num-callers=25 --error-exitcode=1 --child-silent-after-fork=yes --suppressions=../../contrib/valgrind.supp  /labhome/dmitrygla/work_auto/ucx
/test/gtest/gtest --gtest_filter=host/test_ucp_mem_type_alloc_before_init.xfer/0
==29583== Memcheck, a memory error detector
==29583== Copyright (C) 2002-2015, and GNU GPL'd, by Julian Seward et al.
==29583== Using Valgrind-3.12.0 and LibVEX; rerun with -h for copyright info
==29583== Command: /labhome/dmitrygla/work_auto/ucx/test/gtest/gtest --gtest_filter=host/test_ucp_mem_type_alloc_before_init.xfer/0
==29583==
[     INFO ] checking dcx: no
[     INFO ] checking ud: no
[     INFO ] checking udx: no
[     INFO ] checking rc: no
[     INFO ] checking rcx: no
[     INFO ] checking shm_ib: yes
[     INFO ] checking ugni: no
[     INFO ] checking self: yes
[     INFO ] checking tcp: yes
[     INFO ] checking all: yes
[     INFO ] checking all_rcdc: no
[     INFO ] checking self_rcx: yes
[     INFO ] checking rc_x: no
[     INFO ] checking ud_mlx5: no
[     INFO ] checking shm: yes
[     INFO ] checking rc_ud: no
[     INFO ] checking dc_ud: no
[     INFO ] checking no_ud_ud_x: yes
[     INFO ] Using random seed of 23748
[     INFO ]  Setting for valgrind: UCX_IB_RX_QUEUE_LEN=512
[     INFO ]  Setting for valgrind: UCX_IB_RX_BUFS_GROW=512
[     INFO ]  Setting for valgrind: UCX_MM_RX_BUFS_GROW=128
[     INFO ]  Setting for valgrind: UCX_IB_TX_QUEUE_LEN=128
[     INFO ]  Setting for valgrind: UCX_IB_TX_BUFS_GROW=64
[     INFO ]  Setting for valgrind: UCX_RC_TX_CQ_LEN=256
[     INFO ]  Setting for valgrind: UCX_CM_TIMEOUT=600ms
[     INFO ]  Setting for valgrind: UCX_TCP_TX_BUFS_GROW=512
[     INFO ]  Setting for valgrind: UCX_TCP_RX_BUFS_GROW=512
[     INFO ]  Setting for valgrind: UCX_TCP_RX_SEG_SIZE=16k
Note: Google Test filter = host/test_ucp_mem_type_alloc_before_init.xfer/0
[==========] Running 1 test from 1 test case.
[----------] Global test environment set-up.
[----------] 1 test from host/test_ucp_mem_type_alloc_before_init
[ RUN      ] host/test_ucp_mem_type_alloc_before_init.xfer/0
==29583== WARNING: valgrind ignores shmget(shmflg) SHM_HUGETLB
==29583== Syscall param process_vm_writev(lvec[...]) points to uninitialised byte(s)
==29583==    at 0x77F2F8A: process_vm_writev (in /usr/lib64/libc-2.17.so)
==29583==    by 0x8776B43: uct_cma_ep_common_zcopy (cma_ep.c:110)
==29583==    by 0x8776B43: uct_cma_ep_put_zcopy (cma_ep.c:129)
==29583==    by 0x5873BCB: uct_ep_put_zcopy (uct.h:2332)
==29583==    by 0x5873BCB: ucp_rndv_progress_rma_put_zcopy_inner (rndv.c:734)
==29583==    by 0x5873BCB: ucp_rndv_progress_rma_put_zcopy (rndv.c:693)
==29583==    by 0x5878410: ucp_request_try_send (ucp_request.inl:171)
==29583==    by 0x5878410: ucp_request_send (ucp_request.inl:206)
==29583==    by 0x5878410: ucp_rndv_rtr_handler_inner (rndv.c:1002)
==29583==    by 0x5878410: ucp_rndv_rtr_handler (rndv.c:917)
==29583==    by 0x53C1E05: uct_iface_invoke_am (uct_iface.h:628)
==29583==    by 0x53C2532: uct_mm_iface_invoke_am (mm_iface.h:151)
==29583==    by 0x53C2532: uct_mm_iface_process_recv (mm_iface.c:208)
==29583==    by 0x53C29DE: uct_mm_iface_poll_fifo (mm_iface.c:254)
==29583==    by 0x53C2B52: uct_mm_iface_progress (mm_iface.c:278)
==29583==    by 0x583417C: ucs_callbackq_dispatch (callbackq.h:211)
==29583==    by 0x583A8A9: uct_worker_progress (uct.h:2203)
==29583==    by 0x583A8A9: ucp_worker_progress (ucp_worker.c:1892)
==29583==    by 0x792CD0: ucp_test_base::entity::progress(int) (ucp_test.cc:655)
==29583==    by 0x78FB29: ucp_test::progress(int) const (ucp_test.cc:165)
==29583==    by 0x78FF44: ucp_test::wait(void*, int) (ucp_test.cc:222)
==29583==    by 0x6C4163: test_ucp_mem_type_alloc_before_init_xfer_Test::test_body() (test_ucp_mem_type.cc:119)
==29583==    by 0x5367E5: ucs::test_base::run() (test.cc:276)
==29583==    by 0x5369A8: ucs::test_base::TestBodyProxy() (test.cc:302)
==29583==    by 0x6625B5: ucp_test::TestBody() (ucp_test.h:143)
==29583==    by 0x518E93: void testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) (gtest-all.cc:3562)
==29583==    by 0x514071: void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) (gtest-all.cc:3598)
==29583==    by 0x4FB3F0: testing::Test::Run() (gtest-all.cc:3635)
==29583==    by 0x4FBBCB: testing::TestInfo::Run() (gtest-all.cc:3812)
==29583==    by 0x4FC25B: testing::TestCase::Run() (gtest-all.cc:3930)
==29583==    by 0x502A4B: testing::internal::UnitTestImpl::RunAllTests() (gtest-all.cc:5802)
==29583==    by 0x51A271: bool testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) (gtest-all.cc:3562)
==29583==    by 0x514ED3: bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) (gtest-all.cc:3598)
==29583==  Address 0xaa95fd0 is 0 bytes inside a block of size 10,000 alloc'd
==29583==    at 0x4C29BE3: malloc (vg_replace_malloc.c:299)
==29583==    by 0x5274C9: mem_buffer::allocate(unsigned long, ucs_memory_type) (mem_buffer.cc:48)
==29583==    by 0x52832A: mem_buffer::mem_buffer(unsigned long, ucs_memory_type) (mem_buffer.cc:232)
==29583==    by 0x6C532D: test_ucp_mem_type_alloc_before_init::init() (test_ucp_mem_type.cc:88)
==29583==    by 0x53626E: ucs::test_base::SetUpProxy() (test.cc:228)
==29583==    by 0x662575: ucp_test::SetUp() (ucp_test.h:143)
==29583==    by 0x518E93: void testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) (gtest-all.cc:3562)
==29583==    by 0x514071: void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) (gtest-all.cc:3598)
==29583==    by 0x4FB39D: testing::Test::Run() (gtest-all.cc:3630)
==29583==    by 0x4FBBCB: testing::TestInfo::Run() (gtest-all.cc:3812)
==29583==    by 0x4FC25B: testing::TestCase::Run() (gtest-all.cc:3930)
==29583==    by 0x502A4B: testing::internal::UnitTestImpl::RunAllTests() (gtest-all.cc:5802)
==29583==    by 0x51A271: bool testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) (gtest-all.cc:3562)
==29583==    by 0x514ED3: bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) (gtest-all.cc:3598)
==29583==    by 0x501687: testing::UnitTest::Run() (gtest-all.cc:5416)
==29583==    by 0x52364A: RUN_ALL_TESTS() (gtest.h:20059)
==29583==    by 0x523534: main (main.cc:102)
==29583==  Uninitialised value was created by a heap allocation
==29583==    at 0x4C29BE3: malloc (vg_replace_malloc.c:299)
==29583==    by 0x5274C9: mem_buffer::allocate(unsigned long, ucs_memory_type) (mem_buffer.cc:48)
==29583==    by 0x52832A: mem_buffer::mem_buffer(unsigned long, ucs_memory_type) (mem_buffer.cc:232)
==29583==    by 0x6C532D: test_ucp_mem_type_alloc_before_init::init() (test_ucp_mem_type.cc:88)
==29583==    by 0x53626E: ucs::test_base::SetUpProxy() (test.cc:228)
==29583==    by 0x662575: ucp_test::SetUp() (ucp_test.h:143)
==29583==    by 0x518E93: void testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) (gtest-all.cc:3562)
==29583==    by 0x514071: void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) (gtest-all.cc:3598)
==29583==    by 0x4FB39D: testing::Test::Run() (gtest-all.cc:3630)
==29583==    by 0x4FBBCB: testing::TestInfo::Run() (gtest-all.cc:3812)
==29583==    by 0x4FC25B: testing::TestCase::Run() (gtest-all.cc:3930)
==29583==    by 0x502A4B: testing::internal::UnitTestImpl::RunAllTests() (gtest-all.cc:5802)
==29583==    by 0x51A271: bool testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) (gtest-all.cc:3562)
==29583==    by 0x514ED3: bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) (gtest-all.cc:3598)
==29583==    by 0x501687: testing::UnitTest::Run() (gtest-all.cc:5416)
==29583==    by 0x52364A: RUN_ALL_TESTS() (gtest.h:20059)
==29583==    by 0x523534: main (main.cc:102)
==29583==
common/test.cc:261: Failure
Failed
Got 3 valgrind errors during the test
[  FAILED  ] host/test_ucp_mem_type_alloc_before_init.xfer/0, where GetParam() = all (1204 ms)
[----------] 1 test from host/test_ucp_mem_type_alloc_before_init (1222 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test case ran. (1302 ms total)
[  PASSED  ] 0 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] host/test_ucp_mem_type_alloc_before_init.xfer/0, where GetParam() = all

 1 FAILED TEST
==29583==
==29583== HEAP SUMMARY:
==29583==     in use at exit: 9,447 bytes in 39 blocks
==29583==   total heap usage: 166,888 allocs, 166,316 frees, 19,352,489 bytes allocated
==29583==
==29583== LEAK SUMMARY:
==29583==    definitely lost: 0 bytes in 0 blocks
==29583==    indirectly lost: 0 bytes in 0 blocks
==29583==      possibly lost: 0 bytes in 0 blocks
==29583==    still reachable: 9,447 bytes in 39 blocks
==29583==         suppressed: 0 bytes in 0 blocks
==29583== Reachable blocks (those to which a pointer was found) are not shown.
==29583== To see them, rerun with: --leak-check=full --show-leak-kinds=all
==29583==
==29583== For counts of detected and suppressed errors, rerun with: -v
==29583== ERROR SUMMARY: 3 errors from 1 contexts (suppressed: 0 from 0)
make: *** [test_valgrind] Error 1
```

Reproduced on hpc-test-node-upstream02

--- a/test/gtest/ucp/test_ucp_mem_type.cc
+++ b/test/gtest/ucp/test_ucp_mem_type.cc
@@ -96,6 +96,7 @@ public:
         m_recv_buffer.reset();
     }
 
+    static const uint64_t SEED = 0x1111111111111111lu;
 protected:
     size_t                     m_size;
     ucs::auto_ptr<mem_buffer>  m_send_buffer, m_recv_buffer;
@@ -109,7 +110,11 @@ UCS_TEST_P(test_ucp_mem_type_alloc_before_init, xfer) {
     EXPECT_EQ(mem_type(), ucp_memory_type_detect(receiver().ucph(),
                                                  m_recv_buffer->ptr(), m_size));
 
+    mem_buffer::pattern_fill(m_send_buffer->ptr(), m_size, SEED, mem_type());
+
     for (int i = 0; i < 3; ++i) {
+        mem_buffer::pattern_fill(m_recv_buffer->ptr(), m_size, 0, mem_type());
+
         void *sreq = ucp_tag_send_nb(sender().ep(), m_send_buffer->ptr(), m_size,
                                      ucp_dt_make_contig(1), 1,
                                      (ucp_send_callback_t)ucs_empty_function);
@@ -118,6 +123,8 @@ UCS_TEST_P(test_ucp_mem_type_alloc_before_init, xfer) {
                                      (ucp_tag_recv_callback_t)ucs_empty_function);
         wait(sreq);
         wait(rreq);
+
+        mem_buffer::pattern_check(m_recv_buffer->ptr(), m_size, SEED, mem_type());
     }
 }
 

